### ISSUE TITLE
Warn on undefined variable/function in other namespace

### ISSUE THREAD

**Issue by actonDev** [clj-kondo/clj-kondo#634]
**Is your feature request related to a problem? Please describe.**
If in namespace i require another namespace (let's say `(:require [foo])` and i try to refer a variable `foo/var1` (where var1 is not defined) or even try to call `(foo/fn1)` where `fn1` is not defined, clj-kondo doesn't report anything.

# TODO (added by @borkdude)

- [x] I only get 2 failing tests for `goog.global` and `throw` in CLJS, I'll address those soon. The rests of the tests now work again.
- [x] For calls like `(set/onion)` we should now see a warning, but not yet for `(apply set/onion ..)`
- [x] Don't warn more than one time about the same unresolved var
- [x] Unresolved var from clojure.core or cljs.core isn't reported yet
- [x] Name of this linter: `:unresolved-var`, defaults to `:error` level and has config to suppress an entire namespace `foo.bar` or a selection of linters `foo.bar/baz`.

**Comment by borkdude** [comment]
I think this is a duplicate of #546

**Comment by actonDev** [comment]
I saw it but I don't think so (though some similarities)
#546 is for the case  `(:require [foo] :refer [fn1])` not causing a warning when `fn1` is not present.

Or, I guess we could say that this one is more generic, and the #546 is a more special case with the refers.
In any case, you know how to proceed (maybe close this or the other). Was just using clj-kondo and stumbled upon this.
Cheers

**Comment by borkdude** [comment]
@actonDev You're right, this one is more general. I will consider it. One challenge is that people might not have linted the `foo` namespace yet, so it is a little tricky to avoid false positives.

**Comment by mrrodriguez** [comment]
It can be convenient that it ignores cross-ns references in the linting to avoid false positives for cases where macros define symbols that kondo is not aware of.

If a future version of clj-kondo were to attempt to have a linter for these cases, I'd vote for it to be something that could be turned off (perhaps at various levels of granularity) for cases where it gives too many false positives.

**Comment by didibus** [comment]
This would cover the number one use case for why people try to use tools.namespace refresh functionality at the REPL, which is after you've performed a rename of a function/macro/var or you've moved a function/macro/var to a different namespace, or you've deleted a function/macro/var, you want to make sure that nothing in your source code is still depending on the old one, but at the REPL everything will work, so the REPL does not help you here. You either need to restart the REPL (slow), or try to use refresh or a reloaded workflow (which can be difficult and tricky to do right).

To avoid false positives, I'd just vote that if the namespace hasn't been linted yet, then the linter would just consider that all var could exist so it won't warn on using an undefined var unless the linter ran on its namespace and exists in the cache.

> It can be convenient that it ignores cross-ns references in the linting to avoid false positives for cases where macros define symbols that kondo is not aware of

I don't see how this would be a problem here? Macros would not have you refer to a Var in a non existing namespace, they'd just have you use symbols that aren't referred in the current namespace, for which clj-kondo already support you teaching it about those.

**Comment by borkdude** [comment]
@didibus One are where this would cause false positives is with libs like hugsql: it generates a list of vars from a .sql files. But using a config to ignore certain namespaces like `foo.bar.hugsql-vars` this should be possible.

I do think that it's better, certainly initially, to have this linter disabled by default, but we'll see.

**Comment by didibus** [comment]
> One area where this would cause false positives is with libs like hugsql: it generates a list of vars from a .sql files. But using a config to ignore certain namespaces like foo.bar.hugsql-vars this should be possible.

Ya, that's true, things that do dynamic `defs` would be false positives, but I feel things that do that is similar to macros, like they should be teaching clj-kondo somehow of the funky things they do. And in fact, it'll often be a macro that does this. But ya, I can see why due to this the feature is more complicated to implement.

> I do think that it's better, certainly initially, to have this linter disabled by default, but we'll see

I don't know, I feel this will be one of people's favorite linter, Sean even thought clj-kondo already warned for this :p, so people are almost under the assumption it already has this linter turned on haha. And the thing is, if its not made default, libraries will not put the effort to support it, so I think it would just end up with more false positives for the people that turn it on.

I feel there's a distinction between clj-kondo cannot unambiguously distinguish false positives from true positives, like when someone has to say disable the linter, or disable the linter on certain forms or namespaces, and cases where clj-kondo can distinguish the false positives from the true positives, but to do so it needs some additional information provided. Like warnings over a macro that are just because the macro doesn't provide a hook or a lint-as, this isn't a real false positive, its just the macro didn't provide the annotation necessary to clj-kondo to do the linting, but you can actually teach clj-kondo how your macro works, and then it can properly lint without false positives.

I think this feature is similar, it seems like there could be a way for code to properly teach clj-kondo about their dynamic var definitions, and that would allow clj-kondo to properly lint without false positives.

**Comment by borkdude** [comment]
@didibus I have a branch called `unresolved-other-namespace` where this works:

``` clojure
$ clojure -M:clj-kondo --lint - <<< "(require '[clojure.set :as set]) (set/onion #{1 2 3})"
<stdin>:1:34: error: Unresolved symbol: clojure.set/onion
```

Feel free to try it out. I'm getting some false positives still. I would highly appreciate it if you helped me find cases that don't work yet (the tests already give a good overview of this).

**Comment by didibus** [comment]
I started trying it on some code. I found one issue:

```
error: Unresolved symbol: java.lang.RuntimeException/RuntimeException
```

For code:

```
(throw (RuntimeException. msg))
```

I think this is because Clojure imports everything under `java.lang` by default, same way that it requires and refers to `clojure.core` by default.

I had to deal with this in Anakondo, I found the list of all default imports that Clojure uses, you can see it here: https://github.com/didibus/anakondo/blob/master/anakondo.el#L70 its pretty long, they probably all need to be special cased by clj-kondo here.

I'm not sure if different version of Clojure could have more default imports, what I did to find the list of them is just:

```
(ns new-ns)
(ns-imports *ns*)
```

That returns all the imports that Clojure creates automatically in a fresh namespace.

**Comment by didibus** [comment]
Ok, on one of my ClojureScript code base, I get:

```
Syntax error (IllegalAccessError) compiling at (clj_kondo/impl/hooks.clj:1:1).
copy-var does not exist
```

I don't get this error when not using the branch, but instead linting with `clj-kondo v2020.12.12`.

Full error:

```
{:clojure.main/message
 "Syntax error (IllegalAccessError) compiling at (clj_kondo/impl/hooks.clj:1:1).\ncopy-var does not exist\n",
 :clojure.main/triage
 {:clojure.error/phase :compile-syntax-check,
  :clojure.error/line 1,
  :clojure.error/column 1,
  :clojure.error/source "hooks.clj",
  :clojure.error/path "clj_kondo/impl/hooks.clj",
  :clojure.error/class java.lang.IllegalAccessError,
  :clojure.error/cause "copy-var does not exist"},
 :clojure.main/trace
 {:via
  [{:type clojure.lang.Compiler$CompilerException,
    :message
    "Syntax error compiling at (clj_kondo/impl/hooks.clj:1:1).",
    :data
    {:clojure.error/phase :compile-syntax-check,
     :clojure.error/line 1,
     :clojure.error/column 1,
     :clojure.error/source "clj_kondo/impl/hooks.clj"},
    :at [clojure.lang.Compiler load "Compiler.java" 7648]}
   {:type java.lang.IllegalAccessError,
    :message "copy-var does not exist",
    :at [clojure.core$refer invokeStatic "core.clj" 4249]}],
  :trace
  [[clojure.core$refer invokeStatic "core.clj" 4249]
   [clojure.core$refer doInvoke "core.clj" 4217]
   [clojure.lang.RestFn applyTo "RestFn.java" 139]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$load_lib invokeStatic "core.clj" 5966]
   [clojure.core$load_lib doInvoke "core.clj" 5928]
   [clojure.lang.RestFn applyTo "RestFn.java" 142]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$load_libs invokeStatic "core.clj" 5985]
   [clojure.core$load_libs doInvoke "core.clj" 5969]
   [clojure.lang.RestFn applyTo "RestFn.java" 137]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$require invokeStatic "core.clj" 6007]
   [clojure.core$require doInvoke "core.clj" 6007]
   [clojure.lang.RestFn invoke "RestFn.java" 482]
   [clj_kondo.impl.hooks$eval6404$loading__6721__auto____6405
    invoke
    "hooks.clj"
    1]
   [clj_kondo.impl.hooks$eval6404 invokeStatic "hooks.clj" 1]
   [clj_kondo.impl.hooks$eval6404 invoke "hooks.clj" 1]
   [clojure.lang.Compiler eval "Compiler.java" 7177]
   [clojure.lang.Compiler eval "Compiler.java" 7166]
   [clojure.lang.Compiler load "Compiler.java" 7636]
   [clojure.lang.RT loadResourceScript "RT.java" 381]
   [clojure.lang.RT loadResourceScript "RT.java" 372]
   [clojure.lang.RT load "RT.java" 459]
   [clojure.lang.RT load "RT.java" 424]
   [clojure.core$load$fn__6839 invoke "core.clj" 6126]
   [clojure.core$load invokeStatic "core.clj" 6125]
   [clojure.core$load doInvoke "core.clj" 6109]
   [clojure.lang.RestFn invoke "RestFn.java" 408]
   [clojure.core$load_one invokeStatic "core.clj" 5908]
   [clojure.core$load_one invoke "core.clj" 5903]
   [clojure.core$load_lib$fn__6780 invoke "core.clj" 5948]
   [clojure.core$load_lib invokeStatic "core.clj" 5947]
   [clojure.core$load_lib doInvoke "core.clj" 5928]
   [clojure.lang.RestFn applyTo "RestFn.java" 142]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$load_libs invokeStatic "core.clj" 5985]
   [clojure.core$load_libs doInvoke "core.clj" 5969]
   [clojure.lang.RestFn applyTo "RestFn.java" 137]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$require invokeStatic "core.clj" 6007]
   [clojure.core$require doInvoke "core.clj" 6007]
   [clojure.lang.RestFn invoke "RestFn.java" 3894]
   [clj_kondo.impl.analyzer$eval4047$loading__6721__auto____4048
    invoke
    "analyzer.clj"
    1]
   [clj_kondo.impl.analyzer$eval4047 invokeStatic "analyzer.clj" 1]
   [clj_kondo.impl.analyzer$eval4047 invoke "analyzer.clj" 1]
   [clojure.lang.Compiler eval "Compiler.java" 7177]
   [clojure.lang.Compiler eval "Compiler.java" 7166]
   [clojure.lang.Compiler load "Compiler.java" 7636]
   [clojure.lang.RT loadResourceScript "RT.java" 381]
   [clojure.lang.RT loadResourceScript "RT.java" 372]
   [clojure.lang.RT load "RT.java" 459]
   [clojure.lang.RT load "RT.java" 424]
   [clojure.core$load$fn__6839 invoke "core.clj" 6126]
   [clojure.core$load invokeStatic "core.clj" 6125]
   [clojure.core$load doInvoke "core.clj" 6109]
   [clojure.lang.RestFn invoke "RestFn.java" 408]
   [clojure.core$load_one invokeStatic "core.clj" 5908]
   [clojure.core$load_one invoke "core.clj" 5903]
   [clojure.core$load_lib$fn__6780 invoke "core.clj" 5948]
   [clojure.core$load_lib invokeStatic "core.clj" 5947]
   [clojure.core$load_lib doInvoke "core.clj" 5928]
   [clojure.lang.RestFn applyTo "RestFn.java" 142]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$load_libs invokeStatic "core.clj" 5985]
   [clojure.core$load_libs doInvoke "core.clj" 5969]
   [clojure.lang.RestFn applyTo "RestFn.java" 137]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$require invokeStatic "core.clj" 6007]
   [clojure.core$require doInvoke "core.clj" 6007]
   [clojure.lang.RestFn invoke "RestFn.java" 551]
   [clj_kondo.impl.core$eval4039$loading__6721__auto____4040
    invoke
    "core.clj"
    1]
   [clj_kondo.impl.core$eval4039 invokeStatic "core.clj" 1]
   [clj_kondo.impl.core$eval4039 invoke "core.clj" 1]
   [clojure.lang.Compiler eval "Compiler.java" 7177]
   [clojure.lang.Compiler eval "Compiler.java" 7166]
   [clojure.lang.Compiler load "Compiler.java" 7636]
   [clojure.lang.RT loadResourceScript "RT.java" 381]
   [clojure.lang.RT loadResourceScript "RT.java" 372]
   [clojure.lang.RT load "RT.java" 459]
   [clojure.lang.RT load "RT.java" 424]
   [clojure.core$load$fn__6839 invoke "core.clj" 6126]
   [clojure.core$load invokeStatic "core.clj" 6125]
   [clojure.core$load doInvoke "core.clj" 6109]
   [clojure.lang.RestFn invoke "RestFn.java" 408]
   [clojure.core$load_one invokeStatic "core.clj" 5908]
   [clojure.core$load_one invoke "core.clj" 5903]
   [clojure.core$load_lib$fn__6780 invoke "core.clj" 5948]
   [clojure.core$load_lib invokeStatic "core.clj" 5947]
   [clojure.core$load_lib doInvoke "core.clj" 5928]
   [clojure.lang.RestFn applyTo "RestFn.java" 142]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$load_libs invokeStatic "core.clj" 5985]
   [clojure.core$load_libs doInvoke "core.clj" 5969]
   [clojure.lang.RestFn applyTo "RestFn.java" 137]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$require invokeStatic "core.clj" 6007]
   [clojure.core$require doInvoke "core.clj" 6007]
   [clojure.lang.RestFn invoke "RestFn.java" 551]
   [clj_kondo.core$eval146$loading__6721__auto____147
    invoke
    "core.clj"
    1]
   [clj_kondo.core$eval146 invokeStatic "core.clj" 1]
   [clj_kondo.core$eval146 invoke "core.clj" 1]
   [clojure.lang.Compiler eval "Compiler.java" 7177]
   [clojure.lang.Compiler eval "Compiler.java" 7166]
   [clojure.lang.Compiler load "Compiler.java" 7636]
   [clojure.lang.RT loadResourceScript "RT.java" 381]
   [clojure.lang.RT loadResourceScript "RT.java" 372]
   [clojure.lang.RT load "RT.java" 459]
   [clojure.lang.RT load "RT.java" 424]
   [clojure.core$load$fn__6839 invoke "core.clj" 6126]
   [clojure.core$load invokeStatic "core.clj" 6125]
   [clojure.core$load doInvoke "core.clj" 6109]
   [clojure.lang.RestFn invoke "RestFn.java" 408]
   [clojure.core$load_one invokeStatic "core.clj" 5908]
   [clojure.core$load_one invoke "core.clj" 5903]
   [clojure.core$load_lib$fn__6780 invoke "core.clj" 5948]
   [clojure.core$load_lib invokeStatic "core.clj" 5947]
   [clojure.core$load_lib doInvoke "core.clj" 5928]
   [clojure.lang.RestFn applyTo "RestFn.java" 142]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$load_libs invokeStatic "core.clj" 5985]
   [clojure.core$load_libs doInvoke "core.clj" 5969]
   [clojure.lang.RestFn applyTo "RestFn.java" 137]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$require invokeStatic "core.clj" 6007]
   [clojure.core$require doInvoke "core.clj" 6007]
   [clojure.lang.RestFn invoke "RestFn.java" 457]
   [clj_kondo.main$eval140$loading__6721__auto____141
    invoke
    "main.clj"
    1]
   [clj_kondo.main$eval140 invokeStatic "main.clj" 1]
   [clj_kondo.main$eval140 invoke "main.clj" 1]
   [clojure.lang.Compiler eval "Compiler.java" 7177]
   [clojure.lang.Compiler eval "Compiler.java" 7166]
   [clojure.lang.Compiler load "Compiler.java" 7636]
   [clojure.lang.RT loadResourceScript "RT.java" 381]
   [clojure.lang.RT loadResourceScript "RT.java" 372]
   [clojure.lang.RT load "RT.java" 459]
   [clojure.lang.RT load "RT.java" 424]
   [clojure.core$load$fn__6839 invoke "core.clj" 6126]
   [clojure.core$load invokeStatic "core.clj" 6125]
   [clojure.core$load doInvoke "core.clj" 6109]
   [clojure.lang.RestFn invoke "RestFn.java" 408]
   [clojure.core$load_one invokeStatic "core.clj" 5908]
   [clojure.core$load_one invoke "core.clj" 5903]
   [clojure.core$load_lib$fn__6780 invoke "core.clj" 5948]
   [clojure.core$load_lib invokeStatic "core.clj" 5947]
   [clojure.core$load_lib doInvoke "core.clj" 5928]
   [clojure.lang.RestFn applyTo "RestFn.java" 142]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$load_libs invokeStatic "core.clj" 5985]
   [clojure.core$load_libs doInvoke "core.clj" 5969]
   [clojure.lang.RestFn applyTo "RestFn.java" 137]
   [clojure.core$apply invokeStatic "core.clj" 667]
   [clojure.core$require invokeStatic "core.clj" 6007]
   [clojure.main$main_opt invokeStatic "main.clj" 514]
   [clojure.main$main_opt invoke "main.clj" 510]
   [clojure.main$main invokeStatic "main.clj" 664]
   [clojure.main$main doInvoke "main.clj" 616]
   [clojure.lang.RestFn applyTo "RestFn.java" 137]
   [clojure.lang.Var applyTo "Var.java" 705]
   [clojure.main main "main.java" 40]],
  :cause "copy-var does not exist",
  :phase :compile-syntax-check}}
```

**Comment by didibus** [comment]
Some false positive around core.async it looks like:

```
error: Unresolved symbol: clojure.core.async/>!!
error: Unresolved symbol: clojure.core.async/chan
error: Unresolved symbol: clojure.core.async/go-loop
error: Unresolved symbol: clojure.core.async/<!
error: Unresolved symbol: clojure.core.async/go-loop
Unresolved symbol: clojure.core.async/>!
Unresolved symbol: clojure.core.async/close!
Unresolved symbol: clojure.core.async/<!!
```

In all cases the code did something like this:

```
(ns fast-read.core
  (:require [clojure.java.io :as io]
            [clojure.core.async :as a])
  (:gen-class))

(defn process-line
  [file chan]
  (let [r (io/reader file)]
    (reduce
     (fn[_ e] (a/>!! chan e))
     nil
     (line-reducible r))))
```

So its just doing `(ns foo (:require [clojure.core.async :as a]))` and then calling core.async functions with `a/`

**Comment by borkdude** [comment]
@didibus 

> error: Unresolved symbol: java.lang.RuntimeException/RuntimeException

This should now be fixed.

> Syntax error (IllegalAccessError) compiling at (clj_kondo/impl/hooks.clj:1:1).
> copy-var does not exist

It sounds like you are using an older version of sci in that project as well, could this be the case?

The core.async errors should now also be fixed.

I only get 2 failing tests for `goog.global` and `throw` in CLJS, I'll address those soon. The rests of the tests now work again.

The issue is not done yet: for calls like `(set/onion)` we should now see a warning, but not yet for `(apply set/onion ..)`, I'll also address that.

**Comment by borkdude** [comment]
TODO:

- [ ] I only get 2 failing tests for `goog.global` and `throw` in CLJS, I'll address those soon. The rests of the tests now work again.

- [ ] For calls like `(set/onion)` we should now see a warning, but not yet for `(apply set/onion ..)`

**Comment by didibus** [comment]
core.async looks fixed, so does the java.lang.RuntimeException.

Ah yes, I was running it inside my project folder, and `clj` was merging with my project deps when launching clj-kondo, and my project does use `sci`.

Ok, now I've run it so clj-kondo doesn't get merged in with my project deps and I don't get the `hooks` error anymore, but I do get some false positives:

```
error: Unresolved symbol: cljs.core/do

;; For code:
(defonce start (do (mount) true))
```

```
error: Unresolved symbol: cljs.core/try

;; For code:
{:ret (try
              (let [ret (sci/eval-string
                         code
                         {:bindings repl-bindings})]
                {:type (type ret)
                 :ret ret})
              (catch :default e
                {:type :error
                 :err e}))
       :out (str sb)}
```

```
error: Unresolved symbol: cljs.core/set!

;; From Code:
{:did-update
   (fn[state]
     (let [repl-results (rum/deref (::repl-results-ref state))]
       (set! (.-scrollTop repl-results)
             (.-scrollHeight repl-results)))
     state)}
```

```
error: Unresolved symbol: cljs.core/fn*

;; From Code:
(mix/->atom ::code #(-> (first (:rum/args %)) (:init-code)))
```

```
error: Unresolved symbol: cljs.core/if

;; From Code:
(defn ->atom
  [key init]
  {:will-mount
   (fn[state]
     (if (fn? init)
       (assoc state key (atom (init state)))
       (assoc state key (atom init))))})
```

**Comment by borkdude** [comment]
The CLJS false positives are fixed. All tests are now passing again. Keeping a list of TODO in the first post of this issue.

**Comment by borkdude** [comment]
@didibus I merged part 1 to master now. The linter is called `:unresolved-var` since the message you get from `set/onion` is "No such var: set/onion". The default level is `:off` but will be bumped to `:error` when this is done. There are 2 TODO items left in the original post. A test is welcome. You can download a binary from the CircleCI builds if you would like to give this a spin in your daily dev.

**Comment by didibus** [comment]
Could it be somehow it broke the cache lookup? If I run on the command line:

```
clj-kondo --lint .
linting took 32ms, errors: 0, warnings: 0
```

I don't get any errors, but in Emacs (only after having explicitly needed to create the .clj-kondo dir, it didn't work otherwise), I do get:

```
No such var: uut/foo
```

**Comment by borkdude** [comment]
@didibus Should be fixed with `03c556e349a9185abab746f570f95473890c2080`

**Comment by borkdude** [comment]
@didibus Use 1ca9e308ce21ba1078422030ffb30d0f42ee4936 rather than the previous commit, since this contains a fix for another issue that is related to the config and cache.

**Comment by borkdude** [comment]
@didibus Fixed another bug: 13c05a793ce667ef27de9a211129cd43aa4498d2
